Makefile

```diff
@@ -16,75 +15,141 @@ BRIGADE_GITHUB_GW_EXTERNAL_PORT := 7744
-	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$*/rootfs/$* ./$*/cmd/$*
-
-# To use docker-build, you need to have Docker installed and configured. You should also set
-# DOCKER_REGISTRY to your own personal registry if you are not pushing to the official upstream.
-.PHONY: docker-build
-docker-build: build-docker-bins
-docker-build: $(addsuffix -image,$(IMAGES))
-
-%-image:
-	docker build $(DOCKER_BUILD_FLAGS) -t $(DOCKER_REGISTRY)/$*:$(IMAGE_TAG) $*
-
-# You must be logged into DOCKER_REGISTRY before you can push.
-.PHONY: docker-push
-docker-push: $(addsuffix -push,$(IMAGES))
-
-%-push:
-	docker push $(DOCKER_REGISTRY)/$*:$(IMAGE_TAG)
-
-# Cross-compile binaries for our CX targets.
-# Mainly, this is for brig-cross-compile
-%-cross-compile: vendor
-	@for os in $(CX_OSES); do \
-		echo "building $$os"; \
-		for arch in $(CX_ARCHS); do \
-			GOOS=$$os GOARCH=$$arch CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./bin/$*-$$os-$$arch ./$*/cmd/$*; \
-		done; \
-		if [ $$os = 'windows' ]; then \
-			mv ./bin/$*-$$os-$$arch ./bin/$*-$$os-$$arch.exe; \
-		fi; \
-	done
-
-.PHONY: build-release
-build-release: brig-cross-compile
+
+amd64_GOARCH = amd64
+arm64_GOARCH = arm64
+arm32_GOARCH = arm
+
+amd64_QEMU_IMPORT =
+arm64_QEMU_IMPORT = ADD https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/x86_64_qemu-aarch64-static.tar.gz /usr/bin
+arm32_QEMU_IMPORT = ADD https://github.com/multiarch/qemu-user-static/releases/download/v3.1.0-2/x86_64_qemu-arm-static.tar.gz /usr/bin
+
+amd64_DOCKER_ARCH = amd64
+arm64_DOCKER_ARCH = arm64v8
+arm32_DOCKER_ARCH = arm32v6
+
+amd64_DOCKER_ANNOTATIONS = --os linux --arch amd64
+arm64_DOCKER_ANNOTATIONS = --os linux --arch arm64 --variant v8
+arm32_DOCKER_ANNOTATIONS = --os linux --arch arm
+
+.PHONY: build build-images push-images dockerfiles
+
+define GO_CMD_TARGETS
+# Build command
+$1/rootfs/$2/$1: vendor $(shell find $1 -type f -name '*.go')
+	GOOS=linux GOARCH=$$($2_GOARCH) CGO_ENABLED=0 go build -ldflags '$(LDFLAGS)' -o ./$1/rootfs/$2/$1 ./$1/cmd/$1
+
+build-$2 $1-$2-image: $1/rootfs/$2/$1
+endef
+
+define DOCKER_IMAGE_TARGETS
+# Build image for a single arch
+.PHONY: $1-images push-$1-images
+$1-$2-image: $1/Dockerfile.$2
+	docker build $(DOCKER_BUILD_FLAGS) -t $(DOCKER_REGISTRY)/$1:$2-$(IMAGE_TAG) -f $1/Dockerfile.$2 $1
+
+# Push image for a single arch
+push-$1-$2-image: $1-$2-image
+	docker push $(DOCKER_REGISTRY)/$1:$2-$(IMAGE_TAG)
+
+# Create arch dockerfile
+$1/Dockerfile.$2: $1/Dockerfile
+	@cat $1/Dockerfile | ARCH="$2" QEMU_IMPORT="$$($2_QEMU_IMPORT)" DOCKER_ARCH="$$($2_DOCKER_ARCH)" envsubst '$$$${ARCH} $$$${QEMU_IMPORT} $$$${DOCKER_ARCH}' > $1/Dockerfile.$2
+
+dockerfiles: $1/Dockerfile.$2
+build-$2-images $1-images: $1-$2-image
+push-$2-images push-$1-images: push-$1-$2-image
+endef
+
+define DOCKER_MANIFEST_TARGETS
+.PHONY: create-$1-manifest $1
+
+$1: $1-images
+
+create-$1-manifest: push-$1-images
+	# Remove old manifest if present
```

Comment:
Are there any drawbacks in removing the need for this cleanup logic and instead adding `--amend` to the `docker manifest create`?

I explored this option as the cleanup logic broke down for me when my registry was a non-docker.io registry (ACR) -- so the manifest file was named differently (`$(DOCKER_REGISTRY)_$1-$(IMAGE_VERSION)`).  Instead of updating the cleanup logic, I thought it'd be great if we could get rid of it altogether if `docker manifest create --amend ...` would work for us.

Reply:
Honestly, I don't know. The docker manifest commands are so badly documented, that I'm not even sure what they really do. I've just found *a* way to use them that works 😛. My main issue with doing amend would be that you would possibly just elongate the manifest list in repeated builds. This *shouldn't* be a problem as long as the manifest tag and the image tags are the same, but I would have to test it. Also, when doing `manifest push`, we can add  `--purge` (with the option to disable it for debugging purposes), which will purge after pushing.